{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12d466d7-eac9-492a-a7ef-dfbbc2272681",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Import Libraries.\n",
    "\n",
    "import os\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import sparse\n",
    "from scipy.sparse import issparse\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31217ce4-5733-4c7e-a120-d76216a4b089",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load Data.\n",
    "\n",
    "os.chdir(\"/folder/\")\n",
    "adata =  ad.read_h5ad(\"adata.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44db59ed-4bb0-4179-bbbb-b8ee6bb9aa22",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Subset AnnData.\n",
    "\n",
    "adata_subset = adata[adata.obs[\"Status\"].isin([\"C9ALS\", \"sALS\", \"Control\"])].copy()\n",
    "adata_subset =  adata_subset[adata_subset.obs[\"Enrichment\"].isin([\"No\"])].copy()\n",
    "adata_subset =  adata_subset[adata_subset.obs[\"Region\"].isin([\"Spinal_Cord\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1d697a6-7915-4285-a6db-15722a061b1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter genes expressed in ≥20% of cells in ≥1 Status group.\n",
    "\n",
    "groups = adata_subset.obs['Status'].unique()\n",
    "expr_bool = adata_subset.X > 0\n",
    "if not isinstance(expr_bool, np.ndarray):\n",
    "    expr_bool = expr_bool.toarray().astype(bool)\n",
    "\n",
    "keep_genes = []\n",
    "for i, gene in enumerate(adata_subset.var_names):\n",
    "    for group in groups:\n",
    "        cells_in_group = adata_subset.obs['Status'] == group\n",
    "        frac = expr_bool[cells_in_group.values, i].sum() / cells_in_group.sum()\n",
    "        if frac >= 0.2:\n",
    "            keep_genes.append(gene)\n",
    "            break\n",
    "\n",
    "adata_filtered = adata_subset[:, keep_genes].copy()\n",
    "print(f\"Original genes: {adata_subset.n_vars}\")\n",
    "print(f\"After expression filter: {adata_filtered.n_vars}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ce28069-0a21-40d8-a647-0b801e53ae64",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Keep top 80% most expressed genes.\n",
    "\n",
    "counts = adata_filtered.layers['counts_RNA'] if \"counts_RNA\" in adata_filtered.layers else adata_filtered.X\n",
    "mean_expr = np.array(counts.mean(axis=0)).flatten() if issparse(counts) else counts.mean(axis = 0).flatten()\n",
    "\n",
    "percentile_cutoff = 20\n",
    "threshold = np.percentile(mean_expr, percentile_cutoff)\n",
    "genes_to_keep = adata_filtered.var_names[mean_expr >= threshold]\n",
    "adata_filtered_dynamic = adata_filtered[:, genes_to_keep].copy()\n",
    "\n",
    "print(f\"Genes before percentile filter: {adata_filtered.n_vars}\")\n",
    "print(f\"After top 80% filter: {adata_filtered_dynamic.n_vars}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edf96aba-5cdf-4bd7-9641-a4df5aed94ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plot: Fraction of cells expressing each gene.\n",
    "\n",
    "expr_bool = adata_filtered_dynamic.X > 0\n",
    "if not isinstance(expr_bool, np.ndarray):\n",
    "    expr_bool = expr_bool.toarray().astype(bool)\n",
    "frac_expr = expr_bool.sum(axis = 0) / adata_filtered_dynamic.n_obs\n",
    "\n",
    "plt.figure(figsize=(6, 4))\n",
    "plt.hist(frac_expr, bins=30, color='skyblue', edgecolor = 'black')\n",
    "plt.xlabel(\"Fraction of cells expressing gene\")\n",
    "plt.ylabel(\"Number of genes\")\n",
    "plt.title(\"Distribution of gene expression across cells\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4d1f9aa-990b-4d32-9573-c798fb52b2cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plot: Mean expression per gene.\n",
    "\n",
    "counts = adata_filtered_dynamic.layers['counts_RNA'] if \"counts_RNA\" in adata_filtered_dynamic.layers else adata_filtered_dynamic.X\n",
    "mean_expr = np.array(counts.mean(axis = 0)).flatten() if issparse(counts) else counts.mean(axis = 0).flatten()\n",
    "\n",
    "plt.figure(figsize=(6, 4))\n",
    "plt.hist(mean_expr, bins = 30, color = 'lightgreen', edgecolor = 'black')\n",
    "plt.xlabel(\"Mean expression per gene\")\n",
    "plt.ylabel(\"Number of genes\")\n",
    "plt.title(\"Distribution of gene expression levels\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11db1c3c-7f3c-4844-9a71-1f9bef666c60",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create pseudobulk matrix.\n",
    "\n",
    "count_matrix = adata_filtered_dynamic.layers.get(\"counts_RNA\", adata_filtered_dynamic.X).copy()\n",
    "metadata = adata_filtered_dynamic.obs.copy()\n",
    "gene_info = adata_filtered_dynamic.var.copy()\n",
    "\n",
    "if issparse(count_matrix):\n",
    "    count_matrix = count_matrix.toarray()\n",
    "\n",
    "expr_df = pd.DataFrame(count_matrix, index = metadata.index, columns = gene_info.index)\n",
    "expr_df[\"Sample_ID\"] = metadata[\"Sample_ID\"].astype(str)\n",
    "pseudobulk = expr_df.groupby(\"Sample_ID\")[gene_info.index].sum()\n",
    "\n",
    "pseudobulk_metadata = (\n",
    "    metadata.drop_duplicates(subset = [\"Sample_ID\"])\n",
    "    .set_index(\"Sample_ID\")\n",
    "    .loc[pseudobulk.index]\n",
    ")\n",
    "\n",
    "print(\"✅ Pseudobulk shape:\", pseudobulk.shape)\n",
    "print(\"✅ Metadata shape:\", pseudobulk_metadata.shape)\n",
    "\n",
    "for col in [\"Dataset\", \"Sex\", \"Status\"]:\n",
    "    pseudobulk_metadata[col] = pseudobulk_metadata[col].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9dff69a-ede2-4039-aede-aebed2a1b2d4",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Differential Expression (DESeq2)\n",
    "\n",
    "n_cpus = min(20, os.cpu_count())\n",
    "\n",
    "dds = DeseqDataSet(\n",
    "    counts = pseudobulk,\n",
    "    metadata = pseudobulk_metadata,\n",
    "    design_factors = [\"Dataset\", \"Sex\", \"Status\"],\n",
    "    refit_cooks = True,\n",
    "    inference = DefaultInference(n_cpus = n_cpus),\n",
    "    n_cpus = n_cpus,\n",
    ")\n",
    "dds.deseq2()\n",
    "\n",
    "ds = DeseqStats(dds, contrast = [\"Status\", \"sALS\", \"Control\"], inference = DefaultInference(n_cpus = n_cpus)) #Or C9ALS\n",
    "ds.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dd4a019-4b71-4dfc-a4dd-eb1939549458",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_results = ds.results_df.copy()\n",
    "de_results[\"gene\"] = de_results.index\n",
    "de_results.reset_index(drop = True, inplace = True)\n",
    "\n",
    "de_results.to_excel(\"DGE_Analysis.xlsx\", index = False)\n",
    "print(de_results.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "398d9e70-dc9c-426c-a8a7-e38b52cd231a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Matrix Plot.\n",
    "\n",
    "marker_genes = [\"Feature_A\", \"Feature_B\", \"Feature_C\"]\n",
    "\n",
    "status_order = [\"C9ALS\", \"Control\", \"sALS\", \"fALS\"]\n",
    "adata.obs[\"Status\"] = pd.Categorical(adata.obs[\"Status\"], categories = status_order, ordered = True)\n",
    "adata.obs[\"Status_Rank\"] = adata.obs[\"Status\"].cat.codes\n",
    "\n",
    "sample_order = (\n",
    "    adata.obs[[\"Sample_ID\", \"Status_Rank\"]]\n",
    "    .drop_duplicates(\"Sample_ID\")\n",
    "    .sort_values(\"Status_Rank\")[\"Sample_ID\"]\n",
    "    .tolist()\n",
    ")\n",
    "\n",
    "matrix = sc.pl.matrixplot(\n",
    "    adata,\n",
    "    var_names = marker_genes[::-1],\n",
    "    groupby = \"Sample_ID\",\n",
    "    categories_order = sample_order,\n",
    "    layer = \"log1p_normalized\",\n",
    "    cmap = \"Spectral_r\",\n",
    "    standard_scale = \"var\",\n",
    "    colorbar_title = \"Expression\",\n",
    "    title = \"Matrix Plot of Cell Type Marker Genes\",\n",
    "    swap_axes = True,\n",
    "    return_fig = True,\n",
    "    show = False\n",
    ")\n",
    "\n",
    "for key, ax in matrix.get_axes().items():\n",
    "    if isinstance(ax, matplotlib.axes.Axes):\n",
    "        ax.set_facecolor(\"white\")\n",
    "        for spine in ax.spines.values():\n",
    "            spine.set_visible(False)\n",
    "        for label in ax.get_yticklabels():\n",
    "            label.set_fontweight(\"bold\")\n",
    "            label.set_color(\"#3A3A3A\")\n",
    "        ax.set_xlabel(\"Sample ID\", fontsize = 12)\n",
    "        ax.set_ylabel(\"Marker Genes\", fontsize = 12)\n",
    "        ax.tick_params(axis = \"x\", rotation = 90, labelsize = 10)\n",
    "        ax.tick_params(axis = \"y\", labelsize = 10)\n",
    "\n",
    "matrix.fig.set_size_inches(4, 6)\n",
    "matrix.fig.tight_layout()\n",
    "matrix.fig.savefig(\n",
    "    \"matrixplot_marker_genes.png\",\n",
    "    dpi = 800,\n",
    "    bbox_inches = \"tight\",\n",
    "    transparent = True\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pySCENIC",
   "language": "python",
   "name": "pyscenic"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
